File: C++/game2048.cpp
#include <windows.h>
#include <stdio.h>
#include <iostream>
#include <ctime>
using namespace std;

class game2048 {
    private:
        int brd[4][4] = {
            {0, 0, 0, 0},
            {0, 0, 0, 0},
            {0, 0, 0, 0},
            {0, 0, 0, 0},
        };
        int num = 0;
        int score = 0;
        int max = 0;
        int isEnd = 0;
    public:
        void up();
        void down();
        void right();
        void left();
        void draw();
        void start();
        void add();
        int gameover();
};


// 盤面を描画する関数
void game2048::draw() {
    for (int i=0; i<4; i++) {
        for (int j=0; j<4; j++) {
            cout << "\t" << brd[i][j];
        }
        cout << "\t\n";
    }
}


// 初期配置
void game2048::start() {
    std::srand(time(NULL));
    int rand_1 = rand() % 16;
    int rand_2 = rand_1;
    while(rand_1 == rand_2) {
        rand_2 = rand() % 16;
    }
    brd[rand_1 / 4][rand_1 % 4] = 2;
    brd[rand_2 / 4][rand_2 % 4] = 2;

    printf("%d: start   Score: %d   Max: %d\n", num, score, max);
    draw();
}


// 「2」を追加
void game2048::add() {
    int cnt = 0;
    for (int i=0; i<4; i++) {
        for (int j=0; j<4; j++) {
            if (brd[i][j] == 0) {
                cnt++;
            }
        }
    }

    std::srand(time(NULL));
    int add_place = rand() % cnt + 1;
    
    cnt = 0;
    for (int i=0; i<4; i++) {
        for (int j=0; j<4; j++) {
            if (brd[i][j] == 0) {
                cnt++;
                if (cnt == add_place) {
                    brd[i][j] = 2;
                }
            }
        }
    }
}


// ゲームオーバー関数
int game2048::gameover() {
    int cnt = 0;

    // 空きマスを確認
    for (int i=0; i<4; i++) {
        for (int j=0; j<4; j++) {
            if (brd[i][j] == 0) {
                cnt++;
            }
        }
    }


    // 空きマスが無いとき
    if (cnt == 0) {

        // 縦方向に探索
        for (int i=0; i<3; i++) {
            for (int j=0; j<4; j++) {
                if (brd[i][j] == brd[i+1][j]) {
                    cnt++;
                }
            }
        }

        // 横方向に探索
        for (int i=0; i<4; i++) {
            for (int j=0; j<3; j++) {
                if (brd[i][j] == brd[i][j+1]) {
                    cnt++;
                }
            }
        }

        // 空きマスが無い状態かつ上下左右に同じ数字が無いときはゲームオーバー
        if (cnt == 0) {
            printf("\nGameOver\nNum: %d   Score: %d   Max: %d\n", num, score, max);
            return 1;
        }
    }
     
    return 0;
}


// 上方向
void game2048::up() {
    int cnt = 0;
    int move;

    // 隙間を埋める
    while(move>0) {
        move = 0;
        for (int i=0; i<3; i++) {
            for (int j=0; j<4; j++) {        
                if (brd[i][j] == 0 && brd[i+1][j] != 0) {
                    brd[i][j] = brd[i+1][j];
                    brd[i+1][j] = 0;
                    move++;
                    cnt++;
                }
            }
        }
    }

    // 同じ数字を合体
    for (int i=0; i<3; i++) {
        for (int j=0; j<4; j++) {
            if (brd[i][j] == brd[i+1][j]) {
                //printf("%d : %d\n", brd[i][j], brd[i+1][j]);
                brd[i][j] += brd[i+1][j];
                brd[i+1][j] = 0;
                score += brd[i][j];
                if (brd[i][j] > max) {
                    max = brd[i][j];
                }
                cnt++;
            }
        }
    }

    // 入力方向に寄せたとき
    if (cnt > 0) {

        // 合体に寄って生まれた隙間を埋める
        for (int i=0; i<3; i++) {
            for (int j=0; j<4; j++) {        
                if (brd[i][j] == 0 && brd[i+1][j] != 0) {
                    brd[i][j] = brd[i+1][j];
                    brd[i+1][j] = 0;
                }
            }
        }

        num++;
        printf("%d: up   Score: %d   Max: %d\n", num, score, max);
        add();
        draw();

    // タイルが入力方向に全く動かないとき
    } else {
        printf("%d: up - not move\n", num);
    }
    return;
}


// 下方向
void game2048::down() {
    int cnt = 0;
    int move;
    while(move>0) {
        move = 0;
        for (int i=3; i>0; i--) {
            for (int j=0; j<4; j++) {        
                if (brd[i][j] == 0 && brd[i-1][j] != 0) {
                    brd[i][j] = brd[i-1][j];
                    brd[i-1][j] = 0;
                    move++;
                    cnt++;
                }
            }
        }
    }

    for (int i=3; i>0; i--) {
        for (int j=0; j<4; j++) {        
            if (brd[i][j] == brd[i-1][j]) {
                brd[i][j] += brd[i-1][j];
                brd[i-1][j] = 0;
                score += brd[i][j];
                if (brd[i][j] > max) {
                    max = brd[i][j];
                }
                cnt++;
            }
        }
    }

    if (cnt > 0) {
        for (int i=3; i>0; i--) {
            for (int j=0; j<4; j++) {        
                if (brd[i][j] == 0 && brd[i-1][j] != 0) {
                    brd[i][j] = brd[i-1][j];
                    brd[i-1][j] = 0;
                }
            }
        }

        num++;
        printf("%d: down   Score: %d   Max: %d\n", num, score, max);
        add();
        draw();
    } else {
        printf("%d: down - not move\n", num);
    }
    return;
}


// 右方向
void game2048::right() {
    int cnt = 0;
    int move;
    while(move>0) {
        move = 0;
        for (int i=0; i<4; i++) {
            for (int j=3; j>0; j--) {        
                if (brd[i][j] == 0 && brd[i][j-1] != 0) {
                    brd[i][j] = brd[i][j-1];
                    brd[i][j-1] = 0;
                    move++;
                    cnt++;
                }
            }
        }
    }

    for (int i=0; i<4; i++) {
        for (int j=3; j>0; j--) {        
            if (brd[i][j] == brd[i][j-1]) {
                brd[i][j] += brd[i][j-1];
                brd[i][j-1] = 0;
                score += brd[i][j];
                if (brd[i][j] > max) {
                    max = brd[i][j];
                }
                cnt++;
            }
        }
    }

    if (cnt > 0) {    
        for (int i=0; i<4; i++) {
            for (int j=3; j>0; j--) {        
                if (brd[i][j] == 0 && brd[i][j-1] != 0) {
                    brd[i][j] = brd[i][j-1];
                    brd[i][j-1] = 0;
                }
            }
        }

        num++;
        printf("%d: right   Score: %d   Max: %d\n", num, score, max);
        add();
        draw();
    } else {
        printf("%d: right - not move\n", num);
    }
    return;
}


// 左方向
void game2048::left() {
    int cnt = 0;
    int move;
    while(move>0) {
        move = 0;
        for (int i=0; i<4; i++) {
            for (int j=0; j<3; j++) {        
                if (brd[i][j] == 0 && brd[i][j+1] != 0) {
                    brd[i][j] = brd[i][j+1];
                    brd[i][j+1] = 0;
                    move++;
                    cnt++;
                }
            }
        }
    }

    for (int i=0; i<4; i++) {
        for (int j=0; j<3; j++) {        
            if (brd[i][j] == brd[i][j+1]) {
                brd[i][j] += brd[i][j+1];
                brd[i][j+1] = 0;
                score += brd[i][j];
                if (brd[i][j] > max) {
                    max = brd[i][j];
                }
                cnt++;
            }
        }
    }
    
    if (cnt > 0) {
        for (int i=0; i<4; i++) {
            for (int j=0; j<3; j++) {        
                if (brd[i][j] == 0 && brd[i][j+1] != 0) {
                    brd[i][j] = brd[i][j+1];
                    brd[i][j+1] = 0;
                }
            }
        }

        num++;
        printf("%d: left   Score: %d   Max: %d\n", num, score, max);
        add();
        draw();
    } else {
        printf("%d: left - not move\n", num);
    }
    return;
}


int main() {
    BYTE KeyState[256];
    game2048 game;
    int u_flag=0, d_flag=0, r_flag=0, l_flag=0;
    game.start();

    while(1) {
        
        // 上方向
        if (GetKeyState(VK_UP) & 0x80 && u_flag == 0) {
            u_flag = 1;
            game.up();
        } else if ((GetKeyState(VK_UP) & 0x80) == 0 && u_flag == 1) {
            u_flag = 0;
        
        // 下方向
        } else if (GetKeyState(VK_DOWN) & 0x80 && d_flag == 0) {
            d_flag = 1;
            game.down();
        } else if ((GetKeyState(VK_DOWN) & 0x80) == 0 && d_flag == 1) {
            d_flag = 0;

        // 右方向
        } else if (GetKeyState(VK_RIGHT) & 0x80 && r_flag == 0) {
            r_flag = 1;
            game.right();
        } else if ((GetKeyState(VK_RIGHT) & 0x80) == 0 && r_flag == 1) {
            r_flag = 0;

        // 左方向
        } else if (GetKeyState(VK_LEFT) & 0x80 && l_flag == 0) {
            l_flag = 1;
            game.left();
        } else if ((GetKeyState(VK_LEFT) & 0x80) == 0 && l_flag == 1) {
            l_flag = 0;
        
        // エスケープキーで終了
        } else if (GetKeyState(VK_ESCAPE) & 0x80) {
            printf("exit\n");
            break;
        }

        // ゲームオーバーかチェック
        if (game.gameover() == 1) {
            break;
        }

        Sleep(10);
    }

    printf("\nPlease press 'Enter' key.\n\n");
    while(1) {
        if (GetKeyState(VK_RETURN) & 0x80) {
            break;
        }
        Sleep(10);
    }
        
    return 0;
}

File: Python/game2048.py
#coding: utf-8
from msvcrt import getch
import random
import os
import sys

class game2048:
    def __init__(self):
        self.brd = [[0 for j in range(4)] for i in range(4)]
        self.num = 0
        self.score = 0
        self.max = 0
        
        # タイル移動用パラメータ
        self.direction_dic = {
            "up"    :[1, 0, 0, 0, 3, 4, 1, 1],
            "down"  :[-1, 0, 3, 0, 0, 4, -1, 1],
            "right" :[0, -1, 0, 3, 4, 0, 1, -1],
            "left"  :[0, 1, 0, 0, 4, 3, 1, 1]
            }
    
    
    # タイルの初期配置を行う関数
    def start(self):
        start_tile = random.sample(range(16), k=2)
        self.brd[start_tile[0] // 4][start_tile[0] % 4] = 2
        self.brd[start_tile[1] // 4][start_tile[1] % 4] = 2
        self.draw()
    

    # ゲームオーバーを判定する関数
    def gameover(self):
        cnt = 0

        # 空きマスを探索
        for i in range(4):
            for j in range(4):
                if self.brd[i][j] == 0:
                    cnt += 1
        
        # 空きマスが無いとき
        if cnt == 0:

            # 縦方向に探索
            for i in range(3):
                for j in range(4):
                    if self.brd[i][j] == self.brd[i+1][j]:
                        cnt += 1
            
            # 横方向に探索
            for i in range(4):
                for j in range(3):
                    if self.brd[i][j] == self.brd[i][j+1]:
                        cnt += 1
            
            # 空きマスが無い状態かつ上下左右に同じ数字が無いときはゲームオーバー
            if cnt == 0:
                print("\nGameOver\nNum: {}   Score: {}   Max: {}".format(self.num, self.score, self.max))
                exit(0)

    
    # タイル追加関数
    def add(self):
        add_place = 0
        for h in self.brd:
            for w in h:
                if w == 0:
                    add_place += 1
        add_tile = random.randrange(add_place)

        cnt = 0
        for i in range(4):
            for j in range(4):
                if self.brd[i][j] == 0:
                    if cnt == add_tile:
                        self.brd[i][j] = 2
                        return
                    else:
                        cnt += 1


    # 移動関数    
    def move(self, direction):

        # 入力方向に隙間を埋める
        cnt = 0
        move = 1
        while move > 0:
            move = 0
            for i in range(self.direction_dic[direction][2], self.direction_dic[direction][4], self.direction_dic[direction][6]):
                for j in range(self.direction_dic[direction][3], self.direction_dic[direction][5], self.direction_dic[direction][7]):
                    if self.brd[i][j] == 0 and self.brd[i+self.direction_dic[direction][0]][j+self.direction_dic[direction][1]] != 0:
                        self.brd[i][j] = self.brd[i+self.direction_dic[direction][0]][j+self.direction_dic[direction][1]]
                        self.brd[i+self.direction_dic[direction][0]][j+self.direction_dic[direction][1]] = 0
                        move += 1
                        cnt += 1
        
        # 合体
        cnt = self.combine(direction, cnt)

        # 合体によって生まれた隙間を埋める
        if cnt > 0:
            for i in range(self.direction_dic[direction][2], self.direction_dic[direction][4], self.direction_dic[direction][6]):
                for j in range(self.direction_dic[direction][3], self.direction_dic[direction][5], self.direction_dic[direction][7]):
                    if self.brd[i][j] == 0 and self.brd[i+self.direction_dic[direction][0]][j+self.direction_dic[direction][1]] != 0:
                        self.brd[i][j] = self.brd[i+self.direction_dic[direction][0]][j+self.direction_dic[direction][1]]
                        self.brd[i+self.direction_dic[direction][0]][j+self.direction_dic[direction][1]] = 0
            
            self.num += 1
            print("\n{}: {}   Score: {}   Max: {}".format(self.num, direction, self.score, self.max))
            self.add()
            self.draw()
            self.gameover()
        else:
            print("{}: {} - not move".format(self.num, direction))


    # 合体用関数
    def combine(self, direction, cnt):
        for i in range(self.direction_dic[direction][2], self.direction_dic[direction][4], self.direction_dic[direction][6]):
            for j in range(self.direction_dic[direction][3], self.direction_dic[direction][5], self.direction_dic[direction][7]):
                if self.brd[i][j] == self.brd[i+self.direction_dic[direction][0]][j+self.direction_dic[direction][1]]:
                    self.brd[i][j] += self.brd[i+self.direction_dic[direction][0]][j+self.direction_dic[direction][1]]
                    self.brd[i+self.direction_dic[direction][0]][j+self.direction_dic[direction][1]] = 0

                    # Scoreを加算
                    self.score += self.brd[i][j]
                    if self.brd[i][j] > self.max:
                        self.max = self.brd[i][j]

                    cnt += 1

        return cnt

    # 盤面を描画する関数
    def draw(self):
        for i in range(4):
            print("\t{}\t{}\t{}\t{}\t".format(game.brd[i][0], game.brd[i][1], game.brd[i][2], game.brd[i][3]))


if __name__ == "__main__":
    game = game2048()
    game.start()

    while True:
        key = ord(getch())
        if key == 72:
            game.move("up")
        elif key == 80:
            game.move("down")
        elif key == 77:
            game.move("right")
        elif key == 75:
            game.move("left")
        elif key == 27:
            print("exit")
            break
        